vmware-event-router/internal/metrics/server.go

```diff
@@ -41,51 +40,51 @@ type Server struct {
 }
 
 // NewServer returns an initialized metrics server binding to addr
-func NewServer(cfg connection.Config) (*Server, error) {
-	var username, password string
-	var basicAuth bool
-
-	addr := cfg.Address
-	switch cfg.Auth.Method {
-	case "basic_auth":
-		basicAuth = true
-		username = cfg.Auth.Secret["username"]
-		password = cfg.Auth.Secret["password"]
-	case "none":
-		break
-	default:
-		return nil, errors.Errorf("unsupported authentication method for metrics server: %q", cfg.Auth.Method)
+func NewServer(cfg *config.MetricsProviderConfigDefault) (*Server, error) {
+	if cfg == nil {
+		return nil, errors.New("no metrics server configuration found")
 	}
 
 	logger := log.New(os.Stdout, color.Teal("[Metrics Server] "), log.LstdFlags)
+	basicAuth := true
+
+	if cfg.Auth == nil || cfg.Auth.BasicAuth == nil {
+		logger.Print("no credentials found, disabling authentication for metrics server")
+		basicAuth = false
+	}
+
 	mux := http.NewServeMux()
+
 	switch basicAuth {
 	case true:
-		mux.Handle(endpoint, withBasicAuth(logger, expvar.Handler(), username, password))
+		mux.Handle(endpoint, withBasicAuth(logger, expvar.Handler(), cfg.Auth.BasicAuth.Username, cfg.Auth.BasicAuth.Password))
 	default:
 		mux.Handle(endpoint, expvar.Handler())
 	}
 
 	srv := &Server{
 		http: &http.Server{
-			Addr:         addr,
+			Addr:         cfg.BindAddress,
 			Handler:      mux,
 			ReadTimeout:  httpTimeout,
 			WriteTimeout: httpTimeout,
 		},
 		Logger: logger,
 	}
+
 	return srv, nil
 }
 
 // Run starts the metrics server until the context is cancelled or an error
 // occurs. It will collect metrics for the given event streams and processors.
-func (s *Server) Run(ctx context.Context, bindAddr string) error {
+func (s *Server) Run(ctx context.Context) error {
 	errCh := make(chan error, 1)
 	defer close(errCh)
+
 	go func() {
-		addr := fmt.Sprintf("http://%s%s", bindAddr, endpoint)
+		addr := fmt.Sprintf("http://%s%s", s.http.Addr, endpoint)
```

Comment:
This might be a potential problem in case there is trailing slash in `s.http.Addr`. Not sure whether this is prompted by the JSON hints in the YAML?

Reply:
Good catch, will need to clarify this in the docs and provide some safety in the code.